tests/forms/test_permission.py

```diff
@@ -0,0 +1,142 @@
+# -*- coding: utf-8 -*-
+"""Test permission forms."""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from flask_babel import gettext as _
+
+from xl_auth.permission.forms import EditForm, RegisterForm
+
+from ..factories import PermissionFactory
+
+
+def test_edit_form_validate_permission_id_does_not_exist(permission):
+    """Attempt editing permissions with a permission ID that does not exist."""
+    invalid_permission_id = permission.id + 256
+    form = EditForm(invalid_permission_id, user_id=permission.user.id,
+                    collection_id=permission.collection.id)
+
+    assert form.validate() is False
+    assert _('Permission ID "%(permission_id)s" does not exist',
+             permission_id=invalid_permission_id) in form.permission_id.errors
+
+
+def test_register_form_validate_without_user_id(collection):
+    """Attempt registering entry with empty string for user ID."""
+    form = RegisterForm(user_id='', collection_id=collection.id)
+
+    assert form.validate() is False
+    assert _('This field is required.') in form.user_id.errors
+
+
+def test_edit_form_validate_without_user_id(permission):
+    """Attempt editing entry with empty string for user ID."""
+    form = EditForm(permission.id, user_id='', collection_id=permission.collection.id)
+
+    assert form.validate() is False
+    assert _('This field is required.') in form.user_id.errors
+
+
+def test_register_form_validate_without_collection_id(user):
+    """Attempt registering entry with empty string for collection ID."""
+    form = RegisterForm(user_id=user.id, collection_id='')
+
+    assert form.validate() is False
+    assert _('This field is required.') in form.collection_id.errors
+
+
+def test_edit_form_validate_without_collection_id(permission):
+    """Attempt editing entry with empty string for collection ID."""
+    form = EditForm(permission.id, user_id=permission.user.id, collection_id='')
+
+    assert form.validate() is False
+    assert _('This field is required.') in form.collection_id.errors
+
+
+def test_register_form_validate_permission_already_registered(permission):
+    """Attempt registering a (user_id, collection_id) mapping that is already registered."""
+    form = RegisterForm(user_id=permission.user.id, collection_id=permission.collection.id)
+
+    assert form.validate() is False
+    assert _('Permissions for user ID "%(user_id)s" on collection ID "%(collection_id)s" already '
+             'registered', user_id=permission.user.id, collection_id=permission.collection.id
+             ) in form.user_id.errors
+
+
+def test_edit_form_validate_permission_already_registered(permission):
```

Comment:
@kaipoykio About our discussion on "tedious test writing" this morning... This is what it looks like; a dossen new tests for each and every tiny thing added. 😸 

Do you want to give it a try adding BibDB test coverage after py2-3 migration? Here's the current stats for `xl_auth`:
![screen shot 2017-10-03 at 3 45 35 pm](https://user-images.githubusercontent.com/786326/31128436-19e1b462-a852-11e7-8ef3-be657e857989.png)